### PR TITLE
Add custom pickling hooks to LazyXComAccess

### LINKED ISSUE #28146: Dynamic task context fails to be pickled
### Apache Airflow version

2.5.0

### What happened

When I upgrade to 2.5.0, run dynamic task test failed. 

```py
from airflow.decorators import task, dag
import pendulum as pl

@dag(
    dag_id='test-dynamic-tasks', 
    schedule=None,
    start_date=pl.today().add(days=-3),
    tags=['example'])
def test_dynamic_tasks():

    @task.virtualenv(requirements=[])
    def sum_it(values):
        print(values)

    @task.virtualenv(requirements=[])
    def add_one(value):
        return value + 1

    added_values  = add_one.expand(value = [1,2])
    sum_it(added_values)

dag = test_dynamic_tasks()

```

```log
*** Reading local file: /home/andi/airflow/logs/dag_id=test-dynamic-tasks/run_id=manual__2022-12-06T10:07:41.355423+00:00/task_id=sum_it/attempt=1.log
[2022-12-06, 18:07:53 CST] {taskinstance.py:1087} INFO - Dependencies all met for <TaskInstance: test-dynamic-tasks.sum_it manual__2022-12-06T10:07:41.355423+00:00 [queued]>
[2022-12-06, 18:07:53 CST] {taskinstance.py:1087} INFO - Dependencies all met for <TaskInstance: test-dynamic-tasks.sum_it manual__2022-12-06T10:07:41.355423+00:00 [queued]>
[2022-12-06, 18:07:53 CST] {taskinstance.py:1283} INFO - 
--------------------------------------------------------------------------------
[2022-12-06, 18:07:53 CST] {taskinstance.py:1284} INFO - Starting attempt 1 of 1
[2022-12-06, 18:07:53 CST] {taskinstance.py:1285} INFO - 
--------------------------------------------------------------------------------
[2022-12-06, 18:07:53 CST] {taskinstance.py:1304} INFO - Executing <Task(_PythonVirtualenvDecoratedOperator): sum_it> on 2022-12-06 10:07:41.355423+00:00
[2022-12-06, 18:07:53 CST] {standard_task_runner.py:55} INFO - Started process 25873 to run task
[2022-12-06, 18:07:53 CST] {standard_task_runner.py:82} INFO - Running: ['airflow', 'tasks', 'run', 'test-dynamic-tasks', 'sum_it', 'manual__2022-12-06T10:07:41.355423+00:00', '--job-id', '41164', '--raw', '--subdir', 'DAGS_FOLDER/andi/test-dynamic-task.py', '--cfg-path', '/tmp/tmphudvake2']
[2022-12-06, 18:07:53 CST] {standard_task_runner.py:83} INFO - Job 41164: Subtask sum_it
[2022-12-06, 18:07:53 CST] {task_command.py:389} INFO - Running <TaskInstance: test-dynamic-tasks.sum_it manual__2022-12-06T10:07:41.355423+00:00 [running]> on host sh-dataops-airflow.jinde.local
[2022-12-06, 18:07:53 CST] {taskinstance.py:1511} INFO - Exporting the following env vars:
AIRFLOW_CTX_DAG_EMAIL=andi@google.com
AIRFLOW_CTX_DAG_OWNER=andi
AIRFLOW_CTX_DAG_ID=test-dynamic-tasks
AIRFLOW_CTX_TASK_ID=sum_it
AIRFLOW_CTX_EXECUTION_DATE=2022-12-06T10:07:41.355423+00:00
AIRFLOW_CTX_TRY_NUMBER=1
AIRFLOW_CTX_DAG_RUN_ID=manual__2022-12-06T10:07:41.355423+00:00
[2022-12-06, 18:07:53 CST] {process_utils.py:179} INFO - Executing cmd: /home/andi/airflow/venv38/bin/python -m virtualenv /tmp/venv7lc4m6na --system-site-packages
[2022-12-06, 18:07:53 CST] {process_utils.py:183} INFO - Output:
[2022-12-06, 18:07:54 CST] {process_utils.py:187} INFO - created virtual environment CPython3.8.0.final.0-64 in 220ms
[2022-12-06, 18:07:54 CST] {process_utils.py:187} INFO -   creator CPython3Posix(dest=/tmp/venv7lc4m6na, clear=False, no_vcs_ignore=False, global=True)
[2022-12-06, 18:07:54 CST] {process_utils.py:187} INFO -   seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=/home/andi/.local/share/virtualenv)
[2022-12-06, 18:07:54 CST] {process_utils.py:187} INFO -     added seed packages: pip==22.2.1, setuptools==63.2.0, wheel==0.37.1
[2022-12-06, 18:07:54 CST] {process_utils.py:187} INFO -   activators BashActivator,CShellActivator,FishActivator,NushellActivator,PowerShellActivator,PythonActivator
[2022-12-06, 18:07:54 CST] {process_utils.py:179} INFO - Executing cmd: /tmp/venv7lc4m6na/bin/pip install -r /tmp/venv7lc4m6na/requirements.txt
[2022-12-06, 18:07:54 CST] {process_utils.py:183} INFO - Output:
[2022-12-06, 18:07:55 CST] {process_utils.py:187} INFO - Looking in indexes: http://pypi:8081
[2022-12-06, 18:08:00 CST] {process_utils.py:187} INFO - 
[2022-12-06, 18:08:00 CST] {process_utils.py:187} INFO - [notice] A new release of pip available: 22.2.1 -> 22.3.1
[2022-12-06, 18:08:00 CST] {process_utils.py:187} INFO - [notice] To update, run: python -m pip install --upgrade pip
[2022-12-06, 18:08:00 CST] {taskinstance.py:1772} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/home/andi/airflow/venv38/lib/python3.8/site-packages/airflow/decorators/base.py", line 217, in execute
    return_value = super().execute(context)
  File "/home/andi/airflow/venv38/lib/python3.8/site-packages/airflow/operators/python.py", line 356, in execute
    return super().execute(context=serializable_context)
  File "/home/andi/airflow/venv38/lib/python3.8/site-packages/airflow/operators/python.py", line 175, in execute
    return_value = self.execute_callable()
  File "/home/andi/airflow/venv38/lib/python3.8/site-packages/airflow/operators/python.py", line 553, in execute_callable
    return self._execute_python_callable_in_subprocess(python_path, tmp_path)
  File "/home/andi/airflow/venv38/lib/python3.8/site-packages/airflow/operators/python.py", line 397, in _execute_python_callable_in_subprocess
    self._write_args(input_path)
  File "/home/andi/airflow/venv38/lib/python3.8/site-packages/airflow/operators/python.py", line 367, in _write_args
    file.write_bytes(self.pickling_library.dumps({"args": self.op_args, "kwargs": self.op_kwargs}))
_pickle.PicklingError: Can't pickle <class 'sqlalchemy.orm.session.Session'>: it's not the same object as sqlalchemy.orm.session.Session
[2022-12-06, 18:08:00 CST] {taskinstance.py:1322} INFO - Marking task as FAILED. dag_id=test-dynamic-tasks, task_id=sum_it, execution_date=20221206T100741, start_date=20221206T100753, end_date=20221206T100800
[2022-12-06, 18:08:00 CST] {warnings.py:109} WARNING - /home/andi/airflow/venv38/lib/python3.8/site-packages/airflow/utils/email.py:120: RemovedInAirflow3Warning: Fetching SMTP credentials from configuration variables will be deprecated in a future release. Please set credentials using a connection instead.
  send_mime_email(e_from=mail_from, e_to=recipients, mime_msg=msg, conn_id=conn_id, dryrun=dryrun)

[2022-12-06, 18:08:00 CST] {configuration.py:635} WARNING - section/key [smtp/smtp_user] not found in config
[2022-12-06, 18:08:00 CST] {email.py:229} INFO - Email alerting: attempt 1
[2022-12-06, 18:08:01 CST] {email.py:241} INFO - Sent an alert email to ['andi@google.com']
[2022-12-06, 18:08:01 CST] {standard_task_runner.py:100} ERROR - Failed to execute job 41164 for task sum_it (Can't pickle <class 'sqlalchemy.orm.session.Session'>: it's not the same object as sqlalchemy.orm.session.Session; 25873)
[2022-12-06, 18:08:01 CST] {local_task_job.py:159} INFO - Task exited with return code 1
[2022-12-06, 18:08:01 CST] {taskinstance.py:2582} INFO - 0 downstream tasks scheduled from follow-on schedule check
```

### What you think should happen instead

I expect this sample run passed.

### How to reproduce

_No response_

### Operating System

centos 7.9 3.10.0-1160.el7.x86_64

### Versions of Apache Airflow Providers

```
airflow-code-editor==5.2.2
apache-airflow-providers-celery==3.0.0
apache-airflow-providers-microsoft-mssql==3.1.0
apache-airflow-providers-microsoft-psrp==2.0.0
apache-airflow-providers-microsoft-winrm==3.0.0
apache-airflow-providers-mysql==3.0.0
apache-airflow-providers-redis==3.0.0
apache-airflow-providers-samba==4.0.0
apache-airflow-providers-sftp==3.0.0
autopep8==1.6.0
brotlipy==0.7.0
chardet==3.0.4
pip-chill==1.0.1
pyopenssl==19.1.0
pysocks==1.7.1
python-ldap==3.4.2
requests-credssp==2.0.0
swagger-ui-bundle==0.0.9
tqdm==4.51.0
virtualenv==20.16.2
yapf==0.32.0
```

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by uranusjr**
This fixes #28146. [According to random people on the Internet (the best knowledge source)](https://stackoverflow.com/questions/1412787/), this is caused by subtle import statement discrepencies, and we should find and fix the offending import statements, but I think we can probably agree it’s not practical to review all `Session` imports in the Airflow code base.

As an alternative, I tried to fix this by not attempting to pickle any SQLAlchemy objects at all. Unfortunately SQLAlchemy does not really provide a good way to extract information from a Query object (most things are private), so the most barely reasonable way to approach this is to serialize the query into a SQL string, and rebuild a different query from that string to unpickle. Unfortunately it is impossible to use `count()` against such a Query (I don’t know why, SQLAlchemy just crashes with a very unhelpful exception), so we calculate that part eagerly.

### CODE REVIEW

File: airflow/models/xcom.py [hunk 1]
```diff
@@ -723,6 +723,20 @@ def __eq__(self, other: Any) -> bool:
             return all(x == y for x, y in z)
         return NotImplemented
 
+    def __getstate__(self) -> Any:
+        # We don't want to get into the trouble to pickle the entire SQLAlchemy
```
Comment (dstandish) [inline airflow/models/xcom.py]: it's a little unclear whether you mean it's a big effort to do, or it's a problematic thing to do.

File: airflow/models/xcom.py [hunk 2]
```diff
@@ -723,6 +723,20 @@ def __eq__(self, other: Any) -> bool:
...
+        # query object, but there's no good way to extract query information.
+        # The SQL string for DBAPI is the best we can get. THeoratically we can
```
Comment (dstandish) [inline airflow/models/xcom.py]:
```suggestion
        # The SQL string for DBAPI is the best we can get. Theoratically we can
```

**Review by dstandish — APPROVED** [review]

File: airflow/models/xcom.py (hunk 1 above)
Comment (uranusjr) [reply airflow/models/xcom.py]: Ah, the root issue here is that SQLAlchemy objects don’t pickle well out of the box, so the big effort comes from we need to implement _custom logic_ for every thing in a SQLAlchemy query (e.g. filters, order_bys, hints). I’ll add this to the comments to clarify.

Comment (dstandish) [reply airflow/models/xcom.py]: got you ... ok then i would just suggest changing to "go _through the trouble of pickling_ the"

Comment (dstandish) [reply airflow/models/xcom.py]: wait it's "go _to_ the trouble" :) had to google, myself

**Review by dstandish — APPROVED** [review]

File: airflow/models/xcom.py [hunk 3]
```diff
@@ -723,6 +723,21 @@ def __eq__(self, other: Any) -> bool:
             return all(x == y for x, y in z)
         return NotImplemented
 
+    def __getstate__(self) -> Any:
+        # We don't want to go through the trouble of serializing the entire
```
Comment (dstandish) [inline airflow/models/xcom.py]:
```suggestion
        # We don't want to go to the trouble of serializing the entire
```
sorry, i had it wrong earlier... source : https://dictionary.cambridge.org/us/dictionary/english/go-to-the-trouble-to

Comment (uranusjr) [reply airflow/models/xcom.py]: Huh TIL, I’ve been saying _go through the trouble_ all my life as well!

Comment (dstandish) [reply airflow/models/xcom.py]: maybe it is through! what do i know! maybe _to_ is british

Comment (uranusjr) [reply airflow/models/xcom.py]: So from a quick search the idiom is either _go to the trouble of_ or _go through with_ something, according to both Cambridge and Merriam-Webster. It’s quite conceivable people just mix those things up.

Comment (dstandish) [reply airflow/models/xcom.py]: yeah... sounds about right

**Comment by potiuk** [comment]
cc: @uranusjr -> seems that MySQL hits us back again. I am not 100% sure whether this is a blocker to 2.5.1 - but it might be if it turns out that it is going to affect all MySQL users with Dynamic Task mapping (certainly it has potential to impact them).

**Comment by potiuk** [comment]
cc: @pierrejeambrun @ephraimbuddy -> potential blocker for 2.5.1

**Comment by uranusjr** [comment]
Actually I think this is something simpler and unrelated. With SQLite I get

```
INFO - sqlalchemy.exc.ProgrammingError: (sqlite3.ProgrammingError) Incorrect number of bindings supplied. The current statement uses 4, and there are 0 supplied.
INFO - [SQL: SELECT xcom.value
INFO - FROM xcom JOIN dag_run ON xcom.dag_run_id = dag_run.id
INFO - WHERE xcom."key" = ? AND xcom.task_id = ? AND xcom.dag_id = ? AND xcom.run_id = ? ORDER BY xcom.map_index ASC]
```

Looks like a simple SQL error. Please open a new issue with better description.

**Comment by potiuk** [comment]
Isn't that SQL automatically generated by `query.statement.compile` from an ORM mapping ? At least that was my impression what happens here?

**Comment by uranusjr** [comment]
Yes it is. The SQLAlchemy says that should give us a fully compiled SQL, but apparently it does not actually. I don’t know why (or _how_ can we produce a fully compiled SQL).

**Comment by potiuk** [comment]
> Edit: Found it. We need literal_binds. https://docs.sqlalchemy.org/en/14/faq/sqlexpressions.html#rendering-bound-parameters-inline

Yeah. That's cool. Just wonder - is the current form happening often? Or is it only when there is a specific way of using dynamic tasks (was not obvious for me when reading the description/error?). I think answer to that question might determine  whether 2.5.1rc1 is good to release or not.

**Comment by uranusjr** [comment]
Only when you want to pass results of a dynamic task into a subprocess (most commonly virtualenv, Docker, and Kubernetes, I think).

**Comment by potiuk** [comment]
Ah I see. And I guess it was not working before either, so not a regression. Cool. Everything clear now.

**Pull request merged**